0026-an-extensible-db-design.adoc

```diff
@@ -0,0 +1,181 @@
+* all fields (all columns) can be determined only by the key in the table and no other column. E.g. if for every Han_BTC swap we have the same expiries, then this should be an own table. --> update anomaly
+
+==== BCNF
+* any part of a primary key must not have a functional dependency on another column of the table.
+
+==== 4 NF:
+* no multi-valued dependencies. E.g. a customer who bought 2 items should not be in a table twice with the bought item in one column
+
+
+== Approach
+
+. Data collection: Collect all the data to be stored.
+. First model: Create a first model to act upon.
+. Iterations: Iterate on the model to improve and normalize it.
+. Final model
+
+Constraints we considered besides the normalization constraints:
+
+. FK  relationships should not be defined in a parent table (problem of not being able to create the parent if the child does not exist)
+. Certain data of a swap may only be known to a specific role at a later point in time (after creation of the swap). We want to avoid table updates.
+
+
+== Data collection
+
+We identified the following data sources needed:
+
+. Initial data of each swap when being created for a specific role for a specific alpha and beta to protocol where alpha and beta is one of: `HAN`, `HErc20`, `HALight`
+. Additional data added to the swap for a specific role (e.g. secret hash only known to Bob during announcement)
+. Additional data added to the swap's alpha or beta ledger data.
+
+While not necessarily in scope we considered the following as well:
+
+. Swap State (one per swap, aggregated state - record all states with timestamp)
+. Communication State (one per swap - record all states with timestamp)
+. Ledger State (two per swap, record all states with timestamp and additional information if applicable, e.g. transaction ID)
+
+
+== Final model
+
+Instead of starting from scratch, we start iterating over a slightly adapted table layout based on the one proposed by @bonomat
+https://3.basecamp.com/4403044/buckets/16118249/messages/2583518881#__recording_2600021433[in basecamp].
+
+The final model can be found at the end of this document.
+Our reasoning for creating certain table relations is summed up in the sections below.
+
+=== Swaps
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/swap.puml&fmt=svg[Swaps]
+
+Our first challange was depicting `alpha` and `beta` protocols for the swap.
+Our first attempt was to not indicate this in the `swap` table but instead just use foreign key relationships.
+For that to work we would need to indicate in the child table whether it is `alpha` or `beta` (see below in `swap_protocol`).
+In order to reduce the amount of duplicated data we decided to add a separate table `protocol` and a `1 to many` relation ship from `swap` to `protocol`.
+`protocol` is a _static_ table.
+Static means, it contains all available protocols as identifiers.
```

Comment:
That is really a good point and goes partly towards your first question: https://github.com/comit-network/spikes/pull/62/files#r416269116

You are right though, we can easily integrate this into `swap` as we don't load this from cnd anywhere. This was actually a result of the normalization and thinking about it now, this might be one step too far for our needs. 

Side note: 
Given the tendencies of the team to discuss protocol names (😬) it is more efficient to have only a single fk to the protocol in another table instead of having the the protocol names in here. This allows us to only update one entry in `protocol` if we change the name of a protocol vs updating all related entries in `swap`.

Reply:
Yeah, as mentioned above, I think these kind of names should be expressed in identifiers (like table names or Rust structs) and not as runtime data.